packages/secret-rotator/src/secret-rotator.ts

```diff
@@ -0,0 +1,122 @@
+  iamClient: iam_v1.Iam;
+  secretManagerClient: SecretManagerServiceClient;
+
+  constructor(
+    iamClient: iam_v1.Iam,
+    secretManagerClient: SecretManagerServiceClient
+  ) {
+    this.iamClient = iamClient;
+    this.secretManagerClient = secretManagerClient;
+  }
+
+  public async createServiceAccountKey(
+    client: iam_v1.Iam,
+    serviceAccountProjectId: string,
+    serviceAccountEmail: string
+  ) {
+    const name = `projects/${serviceAccountProjectId}/serviceAccounts/${serviceAccountEmail}`;
+    const key = await client.projects.serviceAccounts.keys.create({
+      name,
+      requestBody: {
+        privateKeyType: 'TYPE_GOOGLE_CREDENTIALS_FILE',
+      },
+    });
+    if (!key.data.privateKeyData) {
+      throw new Error('unable to return data');
+    }
+    return Buffer.from(key.data.privateKeyData, 'base64');
+  }
+
+  public async deleteExpiredServiceAccountKeys(
+    client: iam_v1.Iam,
+    serviceAccountProjectId: string,
+    serviceAccountEmail: string
+  ) {
+    const name = `projects/${serviceAccountProjectId}/serviceAccounts/${serviceAccountEmail}`;
+    const keys = await client.projects.serviceAccounts.keys.list({
+      name,
+    });
+    if (keys.data['keys']) {
+      for (const key of keys.data['keys']) {
+        if (
+          key.name &&
+          key.validBeforeTime &&
+          new Date(key.validBeforeTime).getTime() <
+            new Date(Date.now()).getTime()
+        ) {
+          logger.info(
+            `deleting expired keys for ${serviceAccountEmail}: ${key.name}`
+          );
+          await client.projects.serviceAccounts.keys.delete({name: key.name});
+        }
+      }
+    }
+  }
+
+  public async updateSecret(
+    secretsClient: SecretManagerServiceClient,
+    secretManagerProjectId: string,
+    secretName: string,
+    data: Buffer
+  ) {
+    const parent = `projects/${secretManagerProjectId}/secrets/${secretName}`;
+    const [version] = await secretsClient.addSecretVersion({
+      parent,
+      payload: {
+        data,
+      },
+    });
+    return version.name;
+  }
+
+  public async rotateSecret(
+    serviceAccountProjectId: string,
+    serviceAccountEmail: string,
+    secretManagerProjectId: string,
+    secretName: string
+  ) {
+    logger.info(
+      `creating new key for service account: ${serviceAccountEmail} (${serviceAccountProjectId})`
+    );
+
+    const serviceAccountKey = await this.createServiceAccountKey(
+      this.iamClient,
+      serviceAccountProjectId,
+      serviceAccountEmail
+    );
+    logger.info(`updating secret: ${secretName} (${secretManagerProjectId})`);
+
+    const version = await this.updateSecret(
+      this.secretManagerClient,
+      secretManagerProjectId,
+      secretName,
+      serviceAccountKey
+    );
+    logger.info(`updated secret: ${version}`);
+
+    await this.deleteExpiredServiceAccountKeys(
+      this.iamClient,
+      serviceAccountProjectId,
+      serviceAccountEmail
+    );
```

Comment:
Should we delete expired keys first? If we're at the max keys already, we may never get here if creating a service account key fails due to being at the limit.

Reply:
Good catch.